src/CarbonAware.WebApi/src/Controllers/CarbonAwareController.cs

```diff
@@ -278,12 +278,33 @@ public async Task<IActionResult> GetAverageCarbonIntensity([FromQuery, BindRequi
     [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ValidationProblemDetails))]
     [ProducesResponseType(StatusCodes.Status500InternalServerError, Type = typeof(ValidationProblemDetails))]
     [HttpPost("average-carbon-intensity/batch")]
-    public IActionResult GetAverageCarbonIntensityBatch(IEnumerable<CarbonIntensityBatchDTO> requestedCarbonIntensities)
+    public async IAsyncEnumerable<CarbonIntensityDTO> GetAverageCarbonIntensityBatch(IEnumerable<CarbonIntensityBatchDTO> requestedCarbonIntensities)
     {
-        // Dummy result.
         // TODO: implement this controller method after spec is approved.
-        var result = new List<CarbonIntensityDTO>();
-        return Ok(result);
+        using (var activity = Activity.StartActivity())
+        {
+            foreach (var carbonIntensityBatchDTO in requestedCarbonIntensities)
+            {
+                IEnumerable<Location> locationEnumerable = CreateLocationsFromQueryString(new string[] { carbonIntensityBatchDTO.Location });
```

Comment:
Ya I agree, I think I could make a helper function that doesnt the conversion part and use that within the `CreateLocationsFromQueryString` function

Reply:
Tried addressing this but I had a couple questions. Since Enumerable<Locations> is used all the way out, I just pushed the creation of the List of locations to the aggregator instead of the controller, using a "GetSingleLocationOrThrow" function which is very similar to `GetLocationOrThrow` except it contains a single item. Haven't pushed the change yet bc want to run it by someone

Reply:
Is see your point @bderusha, a it sounds like you might be proposing that: **(A)** for those Controller APIs that do **not** require/support Location-Arrays for each start/end time-interval requested -- aka `forecasts/batch`, `average-carbon-intensity`, and `average-carbon-intensity/batch` -- that one could **(B)** "plumb-all-the-way-through" (the Aggregators) to a new additional version of method _ICarbonAwareDataSource.GetCarbonIntensityAsync_, that takes in a _single_ Location (not IEnumerable), and then **(C)** also change/add-to a new "Props" _CarbonAwareConstants_ used by the _CarbonAwareAggregator_ to additionally support both a multiple "Locations" and a single "Location" variations, since some aggregator methods only required 1-Location, and then **(D)** ensure all _DataSources_ support the new method, and so on...

All that to say, I think it should be "out of scope" for this PR..., and considered as an improvement story later... especially since the above is "just one quick idea" of how to address what (I think) you called out

Reply:
after a chat with @bderusha, will open a follow-up clean-up story